scipy/stats/stats.py

```diff
@@ -7469,6 +7470,254 @@ def brunnermunzel(x, y, alternative="two-sided", distribution="t",
+    for a given
+    - quantile (0<`quantile`<1)
+    - confidence level (0<`confidence`<1)
+    - number of samples `n`.
+
+    Returns the largest index of the sample being a valid lower bound,
+    or `None` if there are not enough samples to derive one.
+
+    Used by the public function confint_quantile().
+
+    .. versionadded:: 1.6.0
+    """
+
+    # compute all probabilities from the binomial distribution for
+    # the quantile of interest
+    bd = binom(n, quantile)
+
+    # the lower bound is the last index before the invert survival
+    # function value for the target confidence level
+    lb = bd.isf(confidence) - 1
+
+    if lb < 0:  # isf returns -1 if there are no matching index
+        return None
+    else:
+        return int(lb)
+
+
+def confint_quantile(x, quantile, confidence, type='one-sided'):
+    r"""Compute non-parametric confidence intervals for any quantile.
+
+    This function implements a non-parametric approach to compute
+    confidence intervals for quantiles. The approach is attributed
+    to Thompson [1]_ and later proven to be applicable to any set
+    of i.i.d. samples [2]_. The computation is based on the observation
+    that the probability of a quantile :math:`q` to be larger than any
+    sample :math:`x_m (1\leq m \leq N)` can be computed as
+
+    .. math::
+
+        \mathbb{P}(x_m \leq q) = 1 - \sum_{k=0}^{m-1} \binom{N}{k}
+        q^k(1-q)^{N-k}
+
+    Furthermore, these probabilities are symmetric, which allows to compute
+    both upper and lower bounds from the same computation:
+
+    .. math::
+
+        \mathbb{P}(x_m \leq q) = \mathbb{P}(x_{N-m+1} \geq 1-q).
+
+    The function computes confidence intervals for a given quantile and
+    confidence level, based on `x` which is either a set of samples
+    (one-dimensional array_like) or the number of samples available.
+    The confidence intervals are valid if and only if the samples are i.i.d.
+
+    Both one-sided and two-sided confidence intervals can be obtained
+    (default is one-sided). The function returns two values: either the bounds
+    for the two one-sided confidence intervals, or the lower and upper bounds
+    of a two-sided confidence interval. The return values are either the
+    indexes of the bounds (if `x` is an integer) or sample values
+    (if `x` is the set of samples). `None` is returned when there are not
+    enough samples to compute the desired confidence interval.
+
+    There is no uniqueness of the two-sided confidence interval
+    (see Notes below). Without further assumption on the samples
+    (eg, the nature of the underlying distribution),
+    the one-sided intervals are optimally tight.
+
+    A similar function is available in the QuantileNPCI R package [3]_.
+    The fundation is the same but it computes the confidence interval
+    bounds by doing interpolations between the sample values, which this
+    function uses only sample values as bounds. This results in slightly
```

Comment:
```suggestion
    function uses only sample values as bounds. This results in slightly
```
Which does "This" refer to?

Reply:
"This" refer to the approach proposed here. I would rephrase as follows:
"Thus, `confint_quantile` returns slightly more conservative intervals (i.e., larger)."

It should be noted that the interpolation method implemented in the R package has no mathematical foundations -- as far as I can tell: without knowing the distribution, we cannot assume anything about the sampling probabilities beyond the empirical samples we have.

Reply:
>  I would rephrase as follows

Looks good, but...

> It should be noted...

I'm not sure if we want to reference the R code at all. It didn't influence your implementation, and since it's not license compatible, we wouldn't want anyone to get the impression that it did. Leave it in with the revision you suggested for now, but we should get another opinion.

Reply:
ack